pkg/controllers/federatedresourcequota/federated_resource_quota_enforcement_controller.go

```diff
@@ -0,0 +1,219 @@
+limitations under the License.
+*/
+
+package federatedresourcequota
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/events"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+const (
+	// QuotaEnforcementControllerName is the controller name that will be used when reporting events.
+	QuotaEnforcementControllerName = "federated-resource-quota-enforcement-controller"
+)
+
+// QuotaEnforcementController reflects the status of the namespaced FederatedResourceQuotas.
+// The controller will reconcile in two cases:
+//  1. When a FederatedResourceQuota is CREATED or UPDATED
+//  2. When a ResourceBinding is DELETED
+//
+// When reconciling the controller will update Overall and OverallUsed, the later of which is only
+// updated in response to ResourceBinding deletions. OverallUsed is primarily populated by the RB validation webhook
+type QuotaEnforcementController struct {
+	client.Client // used to operate Work resources.
+	EventRecorder record.EventRecorder
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The SyncController will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *QuotaEnforcementController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("QuotaEnforcementController reconciling %s", req.NamespacedName.String())
+
+	// First try fetching FederatedResourceQuota
+	quota := &policyv1alpha1.FederatedResourceQuota{}
+	if err := c.Get(ctx, req.NamespacedName, quota); err != nil {
+		if apierrors.IsNotFound(err) {
+			// FederatedResourceQuota may no longer exist, in which case we stop processing.
+			return controllerruntime.Result{}, nil
+		}
+
+		// If not FederatedResourceQuota, check if its a deleted ResourceBinding
+		binding := &workv1alpha2.ResourceBinding{}
+		if err := c.Get(ctx, req.NamespacedName, binding); err != nil {
+			if apierrors.IsNotFound(err) {
+				err := c.collectQuotaStatuses(ctx, binding.Namespace)
+				if err != nil {
+					return controllerruntime.Result{}, err
+				}
+			} else {
+				return controllerruntime.Result{}, err
+			}
+		}
```

Comment:
Maybe you misunderstand the way how the controller works.

The object key passed in `Reconcile()` function can only reference to `FederatedResourceQuota`, that is decided by `For(&policyv1alpha1.FederatedResourceQuota{})` when building this controller.

Reply:
Over complicated things here, I can fix to make it more idiomatic to the way controllers work. Thanks :)